pkg/search/search.go

```diff
@@ -0,0 +1,226 @@
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// searchMaxScore suggests that any score higher than this is not considered a match.
+const searchMaxScore = 25
+
+func isNotExist(err error) bool {
+	return os.IsNotExist(errors.Cause(err))
+}
+
+// RepoOptions is the struct used to search, and stores the different options to filter and configure the output
+type RepoOptions struct {
+	Versions     bool
+	Regexp       bool
+	Devel        bool
+	Version      string
+	MaxColWidth  uint
+	RepoFile     string
+	RepoCacheDir string
+	OutputFormat output.Format
+}
+
+// Run searchs and prints the found charts based of the filters
+func (o *RepoOptions) Run(logger log.Logger, args []string) error {
+	wInfo := logio.NewWriter(logger, log.InfoLevel)
+	o.setupSearchedVersion()
+
+	index, err := o.buildIndex()
+	if err != nil {
+		return err
+	}
+
+	var res []*search.Result
+	if len(args) == 0 {
+		res = index.All()
+	} else {
+		q := strings.Join(args, " ")
+		res, err = index.Search(q, searchMaxScore, o.Regexp)
+		if err != nil {
+			return err
+		}
+	}
+
+	search.SortScore(res)
+	data, err := o.applyConstraint(res)
+	if err != nil {
+		return err
+	}
+
+	return o.OutputFormat.Write(wInfo, &repoSearchWriter{data, o.MaxColWidth})
+}
+
+// setupSearchedVersion sets the version to search the chart
+// if no version is set it sets it to any semver version higher than 0.0.0
+func (o *RepoOptions) setupSearchedVersion() {
+	log.Debug("Original chart version: %q", o.Version)
+
+	if o.Version != "" {
+		return
+	}
+
+	if o.Devel { // search for releases and prereleases (alpha, beta, and release candidate releases).
+		log.Debug("setting version to >0.0.0-0")
+		o.Version = ">0.0.0-0"
+	} else { // search only for stable releases, prerelease versions will be skip
+		log.Debug("setting version to >0.0.0")
+		o.Version = ">0.0.0"
+	}
+}
+
+// applyConstraint get a result list and filters it based on the version constraint set
+func (o *RepoOptions) applyConstraint(res []*search.Result) ([]*search.Result, error) {
+	if o.Version == "" {
+		return res, nil
+	}
+
+	constraint, err := semver.NewConstraint(o.Version)
+	if err != nil {
+		return res, errors.Wrap(err, "an invalid version/constraint format")
+	}
+
+	data := res[:0]
+	foundNames := map[string]bool{}
+	for _, r := range res {
+		// if not returning all versions and already have found a result,
+		// you're done!
+		if !o.Versions && foundNames[r.Name] {
+			continue
+		}
+		v, err := semver.NewVersion(r.Chart.Version)
+		if err != nil {
+			continue
+		}
+		if constraint.Check(v) {
+			data = append(data, r)
+			foundNames[r.Name] = true
+		}
+	}
+
+	return data, nil
+}
+
+// buildIndex loads the repos to add them to the index search
+func (o *RepoOptions) buildIndex() (*search.Index, error) {
+	// Load the repositories.yaml
+	rf, err := repo.LoadFile(o.RepoFile)
+	if isNotExist(err) || len(rf.Repositories) == 0 {
+		return nil, errors.New("no repositories configured")
+	}
+
+	i := search.NewIndex()
```

Comment:
`search.NewIndex()` comes from `cmd/helm/search`. I'm a bit un-easy with that. I wonder, should we port it to `hypper/pkg/`, too?

Reply:
I remember reading a doc file from Helm, listing the things covered under backwards compatibility/semver. I think `pkg/` API is covered, as well as  cmd cli UI is covered, but I don't think `cmd/` API is covered though. But I can't find the doc right now, grr.

Reply:
that is a good point, I didnt noticed that.

Im not even sure why they are cramming everything in the cmd part...

Reply:
ah found it:

```
Go libraries inside of pkg/ SHOULD remain backward compatible, though code inside of cmd/ and internal/ may be changed from release to release without notice.```

Reply:
done!